File: ta4j-core/src/test/java/org/ta4j/core/rules/WaitForRuleTest.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.rules;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import org.junit.Before;
import org.junit.Test;
import org.ta4j.core.BarSeries;
import org.ta4j.core.BaseTradingRecord;
import org.ta4j.core.Trade;
import org.ta4j.core.TradingRecord;
import org.ta4j.core.mocks.MockBarSeriesBuilder;

public class WaitForRuleTest {

    private TradingRecord tradingRecord;
    private WaitForRule rule;
    private BarSeries series;

    @Before
    public void setUp() {
        tradingRecord = new BaseTradingRecord();
        series = new MockBarSeriesBuilder().withData(1, 2, 3, 4, 5).build();
    }

    @Test
    public void waitForSinceLastBuyRuleIsSatisfied() {
        // Waits for 3 bars since last buy trade
        rule = new WaitForRule(Trade.TradeType.BUY, 3);

        assertFalse(rule.isSatisfied(0, null));
        assertFalse(rule.isSatisfied(1, tradingRecord));

        tradingRecord.enter(10);
        assertFalse(rule.isSatisfied(10, tradingRecord));
        assertFalse(rule.isSatisfied(11, tradingRecord));
        assertFalse(rule.isSatisfied(12, tradingRecord));
        assertTrue(rule.isSatisfied(13, tradingRecord));
        assertTrue(rule.isSatisfied(14, tradingRecord));

        tradingRecord.exit(15);
        assertTrue(rule.isSatisfied(15, tradingRecord));
        assertTrue(rule.isSatisfied(16, tradingRecord));

        tradingRecord.enter(17);
        assertFalse(rule.isSatisfied(17, tradingRecord));
        assertFalse(rule.isSatisfied(18, tradingRecord));
        assertFalse(rule.isSatisfied(19, tradingRecord));
        assertTrue(rule.isSatisfied(20, tradingRecord));
    }

    @Test
    public void waitForSinceLastSellRuleIsSatisfied() {
        // Waits for 2 bars since last sell trade
        rule = new WaitForRule(Trade.TradeType.SELL, 2);

        assertFalse(rule.isSatisfied(0, null));
        assertFalse(rule.isSatisfied(1, tradingRecord));

        tradingRecord.enter(10);
        assertFalse(rule.isSatisfied(10, tradingRecord));
        assertFalse(rule.isSatisfied(11, tradingRecord));
        assertFalse(rule.isSatisfied(12, tradingRecord));
        assertFalse(rule.isSatisfied(13, tradingRecord));

        tradingRecord.exit(15);
        assertFalse(rule.isSatisfied(15, tradingRecord));
        assertFalse(rule.isSatisfied(16, tradingRecord));
        assertTrue(rule.isSatisfied(17, tradingRecord));

        tradingRecord.enter(17);
        assertTrue(rule.isSatisfied(17, tradingRecord));
        assertTrue(rule.isSatisfied(18, tradingRecord));

        tradingRecord.exit(20);
        assertFalse(rule.isSatisfied(20, tradingRecord));
        assertFalse(rule.isSatisfied(21, tradingRecord));
        assertTrue(rule.isSatisfied(22, tradingRecord));
    }

    @Test
    public void serializeAndDeserialize() {
        rule = new WaitForRule(Trade.TradeType.BUY, 2);
        RuleSerializationRoundTripTestSupport.assertRuleRoundTrips(series, rule);
        RuleSerializationRoundTripTestSupport.assertRuleJsonRoundTrips(series, rule);
    }
}

File: ta4j-core/src/test/kotlin/org/ta4j/core/rules/WaitForRuleTest.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.rules

import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.ta4j.core.BaseTradingRecord
import org.ta4j.core.Trade.TradeType
import org.ta4j.core.TradingRecord

class WaitForRuleTest {
    private var tradingRecord: TradingRecord? = null
    private var rule: WaitForRule? = null
    @Before
    fun setUp() {
        tradingRecord = BaseTradingRecord()
    }

    @Test
    fun waitForSinceLastBuyRuleIsSatisfied() {
        // Waits for 3 bars since last buy trade
        rule = WaitForRule(TradeType.BUY, 3)
        Assert.assertFalse(rule!!.isSatisfied(0, null))
        Assert.assertFalse(rule!!.isSatisfied(1, tradingRecord))
        tradingRecord!!.enter(10)
        Assert.assertFalse(rule!!.isSatisfied(10, tradingRecord))
        Assert.assertFalse(rule!!.isSatisfied(11, tradingRecord))
        Assert.assertFalse(rule!!.isSatisfied(12, tradingRecord))
        Assert.assertTrue(rule!!.isSatisfied(13, tradingRecord))
        Assert.assertTrue(rule!!.isSatisfied(14, tradingRecord))
        tradingRecord!!.exit(15)
        Assert.assertTrue(rule!!.isSatisfied(15, tradingRecord))
        Assert.assertTrue(rule!!.isSatisfied(16, tradingRecord))
        tradingRecord!!.enter(17)
        Assert.assertFalse(rule!!.isSatisfied(17, tradingRecord))
        Assert.assertFalse(rule!!.isSatisfied(18, tradingRecord))
        Assert.assertFalse(rule!!.isSatisfied(19, tradingRecord))
        Assert.assertTrue(rule!!.isSatisfied(20, tradingRecord))
    }

    @Test
    fun waitForSinceLastSellRuleIsSatisfied() {
        // Waits for 2 bars since last sell trade
        rule = WaitForRule(TradeType.SELL, 2)
        Assert.assertFalse(rule!!.isSatisfied(0, null))
        Assert.assertFalse(rule!!.isSatisfied(1, tradingRecord))
        tradingRecord!!.enter(10)
        Assert.assertFalse(rule!!.isSatisfied(10, tradingRecord))
        Assert.assertFalse(rule!!.isSatisfied(11, tradingRecord))
        Assert.assertFalse(rule!!.isSatisfied(12, tradingRecord))
        Assert.assertFalse(rule!!.isSatisfied(13, tradingRecord))
        tradingRecord!!.exit(15)
        Assert.assertFalse(rule!!.isSatisfied(15, tradingRecord))
        Assert.assertFalse(rule!!.isSatisfied(16, tradingRecord))
        Assert.assertTrue(rule!!.isSatisfied(17, tradingRecord))
        tradingRecord!!.enter(17)
        Assert.assertTrue(rule!!.isSatisfied(17, tradingRecord))
        Assert.assertTrue(rule!!.isSatisfied(18, tradingRecord))
        tradingRecord!!.exit(20)
        Assert.assertFalse(rule!!.isSatisfied(20, tradingRecord))
        Assert.assertFalse(rule!!.isSatisfied(21, tradingRecord))
        Assert.assertTrue(rule!!.isSatisfied(22, tradingRecord))
    }
}